_posts/2021-09-08-deep-dive-into-a-debezium-community-connector-scylla-cdc-source-connector.adoc

```diff
@@ -0,0 +1,229 @@
+
+. Too many queries to Scylla - one query per each stream
+. Too many Kafka Connect offsets - one offset per each stream. Storing offsets means the connector can resume from the last saved position after a crash.
+
+To mitigate those problems, we made a decision to group streams on the client side. We chose to group the streams by VNode. This reduced the count from `number_of_nodes * number_of_vnodes_per_node * number_of_shards` to `number_of_nodes * number_of_vnodes_per_node`. In the case of 4-node i3.metal that means a reduction from 73728 to 1024: only 1024 queries to Scylla and 1024 offsets stored on Kafka.
+
+However, we were still uneasy about the number of offsets to be stored on Kafka. When we looked into other connectors, most of them stored only a single offset or at most tens of offsets per replicated table (and as an effect having a limited scalability). 
+
+To understand why storing thousands of streams on Kafka Connect could be a problem, let’s look at how it works under the hood. Each Kafka Connect record created by a source connector contains a key/value offset, for example: key - `my_table`; offset - `25`, which could represent that the connector finished reading 25 rows in `my_table`. Periodically (configured by `offset.flush.interval.ms`), those offsets are flushed to a Kafka topic called `connect-offsets`, as regular Kafka messages.
+
+Unfortunately, Kafka is not a key/value store. When a connector starts up, it must scan all messages on the `connect-offsets` topic to find the one it needs. When it updates a previously saved offset, it just appends the new value to this topic without deleting the previous entry. It’s not a problem with connectors that have only a single offset - when updated every minute, this topic would hold roughly 10,000 messages after a week. However, in the case of the Scylla CDC Source Connector this number could be several orders of magnitude larger!
+
+Fortunately, this issue can be easily mitigated by setting a more aggressive compaction configuration on the `connect-offsets` topic. With the default settings on the Confluent Platform, this topic could grow up to several hundred megabytes after just a few hours (with a Scylla cluster with tens of nodes and very small `offset.flush.interval.ms`). This made the connector startup time slower, as it had to scan the entire offset topic after a start/restart.  By tuning the `segment.bytes`, `segment.ms` or `cleanup.policy`, `retention.ms` we were able to mitigate the problem and significantly reduce the `connect-offsets` topic size. The first two options specify the frequency of the log compaction process. When a segment is compacted, all messages with the same key are reduced to the latest one (the latest offset). Alternatively, setting a shorter retention time (but one that is larger than Scylla’s CDC retention time) proved to be a good option to reduce the offset topic size.
```

Comment:
The reference to Confluent Platform comes a bit out of nowhere. How about: "With a setting of xyz (which is the default setting for instance on Confluent Platform), this topic ..."? Or just "With a setting of xyz, this topic...".